### PR TITLE
[Feature/NifiConnection] Implementation of NifiConnection controller

### PR DESCRIPTION
**Pull request by juldrixx**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| API breaks?     | no
| Deprecations?   | no
| Related tickets | https://github.com/konpyutaika/nifikop/issues/171
| License         | Apache 2.0

### What's in this PR?

New resource to create connection between component (e.g. NifiDataflow) and some tweeks to improve the code reuse.

### Why?

To add new feature to NiFiKop.
To connect 2 dataflows together without manual action.

### Checklist

- [X] Implementation tested
- [X] Error handling code meets the [guideline](docs/error-handling-guide.md)
- [X] Logging code meets the guideline
- [X] User guide and development docs updated (if needed)
- [x] Append changelog with changes

### CODE REVIEW

File: controllers/controller_common.go [hunk 1]
```diff
@@ -166,6 +166,17 @@ func GetUserRefNamespace(ns string, ref v1alpha1.UserReference) string {
 	return userNamespace
 }
 
+// getComponentRefNamespace returns the expected namespace for a Nifi cluster
+// referenced by a user/dataflow CR. It takes the namespace of the CR as the first
+// argument and the reference itself as the second.
+func GetComponentRefNamespace(ns string, ref v1alpha1.ComponentReference) string {
```
Comment (michael81877) [inline controllers/controller_common.go:172]: Could probably use a unit test for this method!

File: plugins/kubectl-nifikop [hunk 2]
```diff
@@ -0,0 +1,197 @@
+#!/usr/bin/env python3
+#
+# Copyright 2019 Orange
```
Comment (michael81877) [inline plugins/kubectl-nifikop:3]: I think this copyright needs tweaked. We should create lint rules to enforce that the license header is the same across all source files.

File: pkg/util/util.go [hunk 3]
```diff
@@ -141,6 +141,32 @@ func ConvertMapStringToMapStringPointer(inputMap map[string]string) map[string]*
 	return result
 }
 
+// StringSliceCompare returns true if the two lists of string are the same
+func StringSliceCompare(list1 []string, list2 []string) bool {
```
Comment (michael81877) [inline pkg/util/util.go:145]: We ought to add a unit test for this!

**Comment by michael81877** [comment]
This is a neat addition and allows you to declaratively compose various ProccessGroups.

Just a conceptual question about this. Can you help me understand when this would be preferred over just connecting 2 process groups in a new process group, version controlling that, and then deploying that new flow via nifikop?

**Comment by juldrixx** [comment]
> This is a neat addition and allows you to declaratively compose various ProccessGroups.
> 
> Just a conceptual question about this. Can you help me understand when this would be preferred over just connecting 2 process groups in a new process group, version controlling that, and then deploying that new flow via nifikop?

When one of the dataflow can reused multiple time for example you could have:
<pre>
A -> B
A -> C
</pre>

The only difference on `A` will be the parameters applied to it and `B` and `C` are totaly different.
So you want to version `A` independently of `B` and `C`.

And you can have the case where `A` is the source of 2 dataflow directly:
<pre>
    -> B
  /
A
  \
    -> C
</pre>

`B` and `C` are 2 different use cases that are alimentated by the same data from `A`. Tomorrow, `C` may disappear but `B` must still works.

**Comment by erdrix** [comment]
> This is a neat addition and allows you to declaratively compose various ProccessGroups.
> 
> Just a conceptual question about this. Can you help me understand when this would be preferred over just connecting 2 process groups in a new process group, version controlling that, and then deploying that new flow via nifikop?

Let's say you have several use cases that consume the same Kafka topic. With your approach, you have to manage a process group that handles the consumption of the Kafka topic and the processing logic for the different use cases (aka different teams), or you create one consumer per use case (unoptimized). 

In terms of ownership, this can be a real pain, with this approach you could say: 
- I have the platform team responsible for deploying the dataflow to bring the kafka messages to the mutualized NiFi cluster.
- Each team is responsible for defining the dataflow responsible for handling those messages.
- In your deployment, you simply deploy all the dataflows and link them together using this functionality.

Another example is that we have something on top of NiFiKop that allows you to deploy a data pipeline by automating cluster and dataflow deployment for preconfigured transformation units (Operator).

In the first version, we had something where one operator (= NiFiCluster + dedicated NiFiDataflow), and to pass data from one operator to another, we used Google Cloud PubSub. In some cases, it can be very expensive to use multiple NiFi clusters with multiple PubSub subjects. With this approach, we can propose to use the `Internal Queue` approach, which means that we no longer use PubSub, but a NiFi connection instead! 

And finally, this will be a great help for https://github.com/konpyutaika/nifikop/issues/43

**Comment by michael81877** [comment]
Thanks for the detailed responses. This enables several things i'd not considered: independent flow versioning in one cluster and it enables attaching validation process groups for integration environments prior to pushing flows to production. 

I'm onboard!

File: plugins/kubectl-nifikop (hunk 2 above)
Comment (michael81877) [reply plugins/kubectl-nifikop]: I raised #173 to address all of the licensing later on. For now this is probably okay.

File: controllers/nifidataflow_controller.go [hunk 4]
```diff
@@ -100,6 +104,7 @@ func (r *NifiDataflowReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	// Check if the cluster reference changed.
 	original := &v1alpha1.NifiDataflow{}
 	current := instance.DeepCopy()
+	patchCurrent := client.MergeFrom(current.DeepCopy())
```
[the remark this answers is not in the document]
Comment (juldrixx) [reply controllers/nifidataflow_controller.go:107]: In fact `current` and `instance` are different because we update the `ClusterRef` on instance: https://github.com/konpyutaika/nifikop/blob/master/controllers/nifidataflow_controller.go#L105

File: plugins/kubectl-nifikop [hunk 5]
```diff
@@ -0,0 +1,197 @@
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (juldrixx) [reply plugins/kubectl-nifikop:1]: I do not know `Rook-Ceph`. How the toolbox pod is used ?

[the remark this answers is not in the document]
Comment (michael81877) [reply plugins/kubectl-nifikop:1]: Rook-Ceph is a persistent volume provisioner for k8s. It comes with an additional deployment called toolbox which is just another pod that gets spun up with all the utilities needed to interact with ceph clusters (the actual storage system). Here's how they deploy it:

https://github.com/rook/rook/blob/v1.10.0/deploy/charts/rook-ceph-cluster/templates/deployment.yaml

But basically, it's just another image with utilities to interact with the system via command line. You `exec` into the pod and run CLI commands to interact with the system. This is a nice alternative to maintaining utilities on your local machine.

So in this case, we can create another image that contains the basic k8s utilities (kubectl) + your `kubectl-nifikop` utility, wrap it in a deployment, and include it as an option to deploy alongside the nifikop operator.

[the remark this answers is not in the document]
Comment (michael81877) [reply plugins/kubectl-nifikop:1]: Here's an example krew plugin where they're basically doing the same sort of thing we're doing here: 

https://github.com/rook/kubectl-rook-ceph

[the remark this answers is not in the document]
Comment (juldrixx) [reply plugins/kubectl-nifikop:1]: Hmm, It's worth digging into and discussing, I think. What do you think, @erdrix?

[the remark this answers is not in the document]
Comment (erdrix) [reply plugins/kubectl-nifikop:1]: This is a very good idea that we could add to this ticket: https://github.com/konpyutaika/nifikop/issues/30 
That way we could provide a docker image and maybe a k8s deployment with the tools for all the manual operations!

[the remark this answers is not in the document]
Comment (michael81877) [reply plugins/kubectl-nifikop:1]: I'll drop a comment on there suggesting a krew plugin and potentially this utility to start. 👍

[the remark this answers is not in the document]
Comment (juldrixx) [reply plugins/kubectl-nifikop]: I have redone the plugin in Golang. Now we just have to manage the build and the manifest generation for krew in the CI.
But I'm not going to lie, the shell scripting and CI/CD are not my best 😄 

I was strongly inspired by the [Datadog's plugin](https://github.com/DataDog/datadog-operator). We can also draw inspiration from them for the shell scripting and CI/CD but I suggest that this be done in another PR 😁

[the remark this answers is not in the document]
Comment (michael81877) [reply plugins/kubectl-nifikop]: seems reasonable to me!

File: api/v1alpha1/nificluster_types.go [hunk 6]
```diff
@@ -27,10 +27,8 @@ const (
 // NifiClusterSpec defines the desired state of NifiCluster
 type NifiClusterSpec struct {
 	// clientType defines if the operator will use basic or tls authentication to query the NiFi cluster.
-	// +kubebuilder:validation:Enum={"tls","basic"}
 	ClientType ClientConfigType `json:"clientType,omitempty"`
 	// type defines if the cluster is internal (i.e manager by the operator) or external.
-	// +kubebuilder:validation:Enum={"external","internal"}
```
Comment (erdrix) [inline api/v1alpha1/nificluster_types.go:33]: Why this change ?

File: api/v1alpha1/nificluster_types.go [hunk 7]
```diff
@@ -27,10 +27,8 @@ const (
 // NifiClusterSpec defines the desired state of NifiCluster
 type NifiClusterSpec struct {
 	// clientType defines if the operator will use basic or tls authentication to query the NiFi cluster.
-	// +kubebuilder:validation:Enum={"tls","basic"}
```
Comment (erdrix) [inline api/v1alpha1/nificluster_types.go:30]: Why this change ?

File: api/v1alpha1/nificluster_types.go [hunk 8]
```diff
@@ -374,7 +372,6 @@ type SSLSecrets struct {
 	// https://cert-manager.io/docs/concepts/issuer/
 	IssuerRef *cmmeta.ObjectReference `json:"issuerRef,omitempty"`
 	// TODO : add vault
-	// +kubebuilder:validation:Enum={"cert-manager","vault"}
```
Comment (erdrix) [inline api/v1alpha1/nificluster_types.go:377]: Why this change ?

File: api/v1alpha1/nifidataflow_types.go [hunk 9]
```diff
@@ -22,7 +22,6 @@ type NifiDataflowSpec struct {
 	// contains the reference to the ParameterContext with the one the dataflow is linked.
 	ParameterContextRef *ParameterContextReference `json:"parameterContextRef,omitempty"`
 	// if the flow will be synchronized once, continuously or never
-	// +kubebuilder:validation:Enum={"never","always","once"}
```
Comment (erdrix) [inline api/v1alpha1/nifidataflow_types.go:25]: Why this change ?

File: api/v1alpha1/nifidataflow_types.go [hunk 10]
```diff
@@ -33,8 +32,7 @@ type NifiDataflowSpec struct {
 	// contains the reference to the NifiRegistry with the one the dataflow is linked.
 	RegistryClientRef *RegistryClientReference `json:"registryClientRef,omitempty"`
 	// describes the way the operator will deal with data when a dataflow will be updated : drop or drain
-	// +kubebuilder:validation:Enum={"drop","drain"}
```
Comment (erdrix) [inline api/v1alpha1/nifidataflow_types.go:36]: Why this change ?

File: config/rbac/role.yaml [hunk 11]
```diff
@@ -142,6 +142,32 @@ rules:
   - get
   - patch
   - update
+- apiGroups:
```
Comment (erdrix) [inline config/rbac/role.yaml:145]: New resources must also be added to the Helm chart's  RBAC!

File: config/samples/nifi_v1alpha1_nificonnection.yaml [hunk 12]
```diff
@@ -0,0 +1,6 @@
+apiVersion: nifi.konpyutaika.com/v1alpha1
+kind: NifiConnection
+metadata:
+  name: nificonnection-sample
+spec:
+  # TODO(user): Add fields here
```
Comment (erdrix) [inline config/samples/nifi_v1alpha1_nificonnection.yaml:6]: Could you add a complete and documented example?

**Review by erdrix — CHANGES_REQUESTED** [review]
It would be interesting to add documentation on how to use this feature with concrete examples.

File: api/v1alpha1/nificluster_types.go (hunk 8 above)
Comment (juldrixx) [reply api/v1alpha1/nificluster_types.go:377]: To make this kubebuilder annotation available to all the resources using the PKIBackend type to avoid having to copy and paste.

File: api/v1alpha1/nifidataflow_types.go (hunk 10 above)
Comment (juldrixx) [reply api/v1alpha1/nifidataflow_types.go:36]: To make this kubebuilder annotation available to all the resources using the PKIBackend type to avoid having to copy and paste.

File: api/v1alpha1/nifidataflow_types.go (hunk 9 above)
Comment (juldrixx) [reply api/v1alpha1/nifidataflow_types.go:25]: To make this kubebuilder annotation available to all the resources using the DataflowSyncMode type to avoid having to copy and paste.

File: api/v1alpha1/nificluster_types.go (hunk 7 above)
Comment (juldrixx) [reply api/v1alpha1/nificluster_types.go:30]: To make this kubebuilder annotation available to all the resources using the ClientConfigType type to avoid having to copy and paste.

File: api/v1alpha1/nificluster_types.go (hunk 6 above)
Comment (juldrixx) [reply api/v1alpha1/nificluster_types.go:33]: To make this kubebuilder annotation available to all the resources using the ClusterType type to avoid having to copy and paste.

File: config/samples/nifi_v1alpha1_nificonnection.yaml (hunk 12 above)
Comment (juldrixx) [reply config/samples/nifi_v1alpha1_nificonnection.yaml]: done

File: CHANGELOG.md [hunk 13]
```diff
@@ -2,6 +2,7 @@
 
 ### Added
 
+- [PR #168](https://github.com/konpyutaika/nifikop/pull/168) - **[Operator/NifiConnection]** Implementation on NifiConnection controller.
```
[the remark this answers is not in the document]
Comment (juldrixx) [reply CHANGELOG.md:6]: done

File: config/rbac/role.yaml (hunk 11 above)
Comment (juldrixx) [reply config/rbac/role.yaml:145]: done

**Comment by juldrixx** [comment]
> It would be interesting to add documentation on how to use this feature with concrete examples.

What do you mean? I did add some documentation on this subject.
